### PR TITLE
READY: Display quantitative information when hovering over nodes and links

### PR DESCRIPTION
**Pull request by rkeulemans**
Solves #171 

Note: this branch is setup with the example dataset for the sake of easy development. This will be reverted before merge.

Current status (using a table now):
info_new
edge_info

### CODE REVIEW

**Comment by rkeulemans** [comment]
@TimSpeelman @vandenheuvel What do you think of the label format? Should I use a small html table for this or is it fine this way?

**Comment by vandenheuvel** [comment]
A table would be a good idea, this looks somewhat messy. I don't think adding icons and/or emoji's is a good idea, I'm afraid it will clutter the view. Let's stick with displaying basic info only for this PR.

**Comment by MaxVanDeursen** [comment]
I agree with @vandenheuvel, I would suggest aligning the columns and rows i.e. a table. If this hover is already HTML, I don't think it would be too much more work to get this table in in this PR.

**Comment by vandenheuvel** [comment]
Great, this small change gives it a cleaner look. Could you show some more characters from the public key?

**Comment by rkeulemans** [comment]
It is 5 now, can be adjusted in the style_contig.js

**Comment by vandenheuvel** [comment]
I'm assuming that the actual public keys in the database are only one or two characters long?

**Comment by rkeulemans** [comment]
No, but this example dataset has only numbers under 100 in it, the last 5 are shown. The pk of the node in the picture is 0. So ...0 is shown. This looks odd but will be perfect with the real data (128/256 character keys).

File: TriblerGUI/widgets/trustpage/js/request_data.js [hunk 1]
```diff
@@ -49,7 +49,8 @@ function make_cors_request(method, url) {
  * @param callback The callback which is called with the GraphResponseData
  */
 function get_node_info(public_key, callback) {
-    var url = "http://localhost:8085/display?focus_node=" + public_key + "&neighbor_level=1";
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/request_data.js]: Please remove this debugging functionality

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 2]
```diff
@@ -12,6 +12,7 @@
 function onNewData(data) {
     state.request_pending = false;
     update(processData(data));
+    console.log(processData(data));
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/main.js]: Please remove this debugging functionality

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 3]
```diff
@@ -119,21 +171,63 @@ function drawLinks(svg, data) {
...
 /**
+ * Show the link information when the user hovers over a link
+ * @param link the link to get the information of
+ * @param infoLabel the label to display the information in
+ */
+function mouseInEdge(link, infoLabel) {
+    infoLabel.style("opacity", config.node.hoverLabel.opacityOnEdge);
+    infoLabel.style("display", "block");
+
+    infoLabel.html(
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I'd suggest removing some of the newlines in this HTML to improve readability

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 4]
```diff
@@ -28,48 +30,99 @@ function drawNodes(svg, data, on_click) {
...
+ * Show the node information when the user hovers over a node
+ * @param node the node to get the information of
+ * @param infoLabel the label to display the information in
+ */
+function mouseInNode(node, infoLabel) {
+    infoLabel.style("opacity", 1);
+    infoLabel.style("display", "block");
+
+    infoLabel.html(
+        "<table>" +
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I'd suggest removing some of the newlines in this HTML to improve readability

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 5]
```diff
@@ -119,21 +171,63 @@ function drawLinks(svg, data) {
...
+ * @param infoLabel the label to display the information in
+ */
+function mouseInEdge(link, infoLabel) {
+    infoLabel.style("opacity", config.node.hoverLabel.opacityOnEdge);
+    infoLabel.style("display", "block");
+
+    infoLabel.html(
+        "<table>" +
+        "<tr>" +
+        "<td class=\"metric\">Uploaded</td>" +
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I'd suggest renaming Uploaded and Downloaded by Uploaded by X and Uploaded by Y. This makes it independent from which node it is viewed and removes ambiguity

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 6]
```diff
@@ -28,48 +30,99 @@ function drawNodes(svg, data, on_click) {
...
         .style("font-size", config.node.publicKeyLabel.fontSize)
+        .style("font-weight", config.node.publicKeyLabel.fontWeight)
         .style("fill", config.node.publicKeyLabel.color)
         .text(function (d) {
-            return d.public_key.substr(-5);
+            return d.public_key.substr(-3);
+        });
+
+    // Transparent circle for the mouse events
+    groups
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I assume this is done so that:

- The event triggers do not have to be duplicated
- There's no way the text within the nodes is selected and used

Is there no other way to get this functionality without pasting a transparent circle above it?

Comment (rkeulemans) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Assumptions correct. I cannot think of a better solution than this.

Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I do not know whether it is possible, but is there no such way to apply certain attributes to all appended objects? This could make it so that the duplication is not required. I'd also think that there is some way to make text not selectable.

Comment (rkeulemans) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: I think that requires the same amount or even more code.
@TimSpeelman what is your opinion?

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 7]
```diff
@@ -118,22 +169,59 @@ function drawLinks(svg, data) {
 
     links.append("line")
         .attr("class", "link-source")
-        .attr("stroke-width", function (d) {
-            return getStrokeWidth(d, data)
-        })
-        .style("stroke", config.link.colorLinkSource);
+        .style("stroke", config.link.colorLinkSource)
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Add semicolon here

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 8]
```diff
@@ -28,48 +28,100 @@ function drawNodes(svg, data, on_click) {
...
+ */
+function mouseInNode(node, infoLabel) {
+    infoLabel
+        .style("opacity", config.node.hoverLabel.opacityOnNode)
+        .style("display", "block")
+
+        // Update the label with the information corresponding to the node
+        .html(
+        "<table>" +
+            "<tr>" +
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: There is a lot of duplicate code here, you should create a loop.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 9]
```diff
@@ -118,22 +169,59 @@ function drawLinks(svg, data) {
...
+ * @param infoLabel the label to display the information in
+ */
+function mouseInEdge(link, infoLabel) {
+    infoLabel
+        .style("opacity", config.node.hoverLabel.opacityOnEdge)
+        .style("display", "block")
+
+        // Update the label with the information corresponding to the link
+        .html(
+            "<table>" +
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Similar to above, too much duplicate code.

File: TriblerGUI/widgets/trustpage/style.css [hunk 10]
```diff
@@ -7,11 +7,26 @@ html, body {
     background-color: #202020;
 }
 
+.metric {
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/style.css]: I personally think this name is somewhat unclear, shouldn't this at least be plural?

**Comment by rkeulemans** [comment]
Does making a table of 4 rows really need a loop? This will increase code complexity and understandably because of a more complex mechanism for creating the table. Did you consider this?

**Comment by vandenheuvel** [comment]
Yes I did consider that. It doesn't necessarily increase complexity and ease of understanding. This should be quite easy especially with Python. It really does need a loop.

**Comment by rkeulemans** [comment]
Will implement it then. Did you notice this is JavaScript?

**Comment by vandenheuvel** [comment]
Oh wow.

**Comment by rkeulemans** [comment]
Hope you like it. Cannot test it with the sample dataset because somebody screwed the comment in the master explaining how to do this.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 11]
```diff
@@ -28,48 +28,102 @@ function drawNodes(svg, data, on_click) {
     // Always remove existing nodes before adding new ones
     selectNodes(svg).remove();
 
+    // Do not display the label after the graph is updated
+    hoverInfoLabel.style("display", "none");
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: This does not belong here, please move to main.js right after the call to drawNodes()

**Comment by vandenheuvel** [comment]
@rkeulemans I was thinking more about something like this:
```
rowItems = {"key1": 1,
          "key2": 2};

html = "<table>";
for (var key in rowItems) {
    html += "<tr>" +
                "<td class=\"metric\">" + key + "</td>" +
                "<td>" + rowItems[key] + "</td>" +
            "</tr>";
}
html += "</table>";
```

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 12]
```diff
@@ -28,48 +28,102 @@ function drawNodes(svg, data, on_click) {
...
+        .append("circle")
+        .style("fill-opacity", "0")
+        .attr("r", config.node.circle.radius)
+        .attr("cx", config.node.circle.cx)
+        .attr("cy", config.node.circle.cy)
+        .style("cursor", config.node.circle.cursor)
+        .on("click", on_click)
+        .on("mouseover", function (d) {
+            mouseInNode(d, hoverInfoLabel);
+        })
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: since you need hoverInfoLabel here, pass it as a parameter to drawNodes, because now you are simply assuming that it is defined somewhere else.

**Comment by rkeulemans** [comment]
And this is just the format I think is not clear. The helper functions make it easy to follow what is going on. This would need additional comments and it is still not instantly clear what is happening without reading them.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 13]
```diff
@@ -28,48 +28,102 @@ function drawNodes(svg, data, on_click) {
...
+        tableRows.concat(createTableRow(quantity.key, quantity.value));
+    return tableRows;
+}
+
+function createTableRow(quantityDescription, quantityValue) {
+    return "<tr>" +
+        "<td class=\"metric\"> + quantityDescription + </td>" +
+        "<td> + quantityValue + </td>" +
+        "</tr>";
+}
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: This is a bit inconsistent with the above. You can use d3 for creating html. Something like this, but then more:
```
    var rows = d3.select("table").selectAll("tr").data(quantities)
    .exit().remove() // this removes all old table rows
    .enter().append('tr') // this makes new table rows for the new data
    rows.append('td') // append a cell
    .attr('class','metric')
    .html(function(d){return d.key})
```

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 14]
```diff
@@ -118,22 +171,59 @@ function drawLinks(svg, data) {
...
+
+    d3.selectAll(".link")
         .attr("stroke-width", function (d) {
-            return getStrokeWidth(d, data)
+            return getStrokeWidth(d, data);
         })
-        .style("stroke", config.link.colorLinkTarget);
+        .on("mouseover", function (d) {
+            mouseInEdge(d, hoverInfoLabel);
+        })
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: please pass as function parameter.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 15]
```diff
@@ -118,22 +171,59 @@ function drawLinks(svg, data) {
...
+        .on("mouseout", function () {
+            hoverInfoLabel.style("opacity", 0);
+        });
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: please pass as function parameter.

File: TriblerGUI/widgets/trustpage/style.css (hunk 10 above)
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/style.css]: no not plural since the usage corresponds to a single metric. when this class would apply to the table itself then .metrics makes sense

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 16]
```diff
@@ -121,6 +109,20 @@ function mouseInNode(node, infoLabel) {
...
+function getTableContents(quantities) {
+    var tableRows = "";
+    for (var quantity in quantities)
+        tableRows.concat(createTableRow(quantity.key, quantity.value));
+    return tableRows;
+}
+
+function createTableRow(quantityDescription, quantityValue) {
+    return "<tr>" +
+        "<td class=\"metric\"> + quantityDescription + </td>" +
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: @rkeulemans I believe there are some double quote's missing around `quantityDescription` and `quantityValue`.

File: TriblerGUI/widgets/trustpage/js/request_data.js [hunk 17]
```diff
@@ -49,7 +49,8 @@ function make_cors_request(method, url) {
  * @param callback The callback which is called with the GraphResponseData
  */
 function get_node_info(public_key, callback) {
-    var url = "http://localhost:8085/multichain/network?focus_node=" + public_key + "&neighbor_level=1";
+    // var url = "http://localhost:8085/multichain/network?focus_node=" + public_key + "&neighbor_level=1";
+    var url = "http://localhost:8085/multichain/network?dataset=static&focus_node=10&neighbor_level=1"
```
Comment (rkeulemans) [inline TriblerGUI/widgets/trustpage/js/request_data.js]: Will fix this!

**Comment by rkeulemans** [comment]
Since massive master merge [57f2935](https://github.com/vandenheuvel/tribler/commit/57f2935a179197cd61bfa7ea6d7bbdbeadc52e21) after rebasing the code is broken see [information_on_hover_rebase] (https://github.com/vandenheuvel/tribler/commit/e5865d6f10435a88eb0527e5ec41b0e31bcc8d71).

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 18]
```diff
@@ -130,23 +136,28 @@ function update(graph) {
...
             return node.graphNode.public_key === state.previous_focus_pk;
         });
-        var correction = target_angle - previous_focus.alpha;
-
-        state.tree.nodes.forEach(function (node) {
-            node.alpha += correction;
-        });
+        if (previous_focus) {
+            var correction = target_angle - previous_focus.alpha;
+            console.log('correction', correction);
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/trustpage/js/main.js]: Remove debug functionality.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 19]
```diff
@@ -133,24 +191,35 @@ function drawLinks(svg, data) {
         .attr("class", "link")
         .style("opacity", "0");
 
+    // A link between two nodes is composed of two parts, representing the amount of data uploaded and downloaded
+
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Please move the above explanation to the function description

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 20]
```diff
@@ -41,45 +42,102 @@ function drawNodes(svg, data, on_click) {
...
         .attr("r", "0")
         .attr("cx", config.node.circle.cx)
-        .attr("cy", config.node.circle.cy)
-        .style("cursor", config.node.circle.cursor)
-        .on("click", on_click)
-        .on("mouseenter", function () {
-            d3.select(this).transition().ease(d3.easeElasticOut).delay(0).duration(300).attr("r", 25);
-        }).on("mouseout", function () {
-            d3.select(this).transition().ease(d3.easeElasticOut).delay(0).duration(300).attr("r", 20);
-        });
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Why are these lines removed?

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 21]
```diff
@@ -133,24 +191,35 @@ function drawLinks(svg, data) {
...
-            return getStrokeWidth(d, data)
-        })
         .style("stroke", config.link.colorLinkTarget);
 
-    links.append("line")
-        .attr("class", "link-velocity")
-        .attr("stroke-width", 2)
-        .style("stroke", "white");
+    // Functionality for the composed link
+    d3.selectAll(".link")
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Please change this to ```links```.

Now, everytime you call ```drawLinks```, it adds .on(mouseover) etc. to all links, instead of just the new ones. The old ones now have two exactly the same event listeners (the callback is executed twice). This keeps increasing when you click new nodes.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 22]
```diff
@@ -160,17 +229,36 @@ function drawLinks(svg, data) {
...
 /**
+ * Show the link information when the mouse enters a link
+ * @param linkData the link to get the information of
+ * @param hoverInfoLabel the label to display the information in
+ */
+function mouseOverLink(linkData, hoverInfoLabel) {
+    // The quantity descriptions and the corresponding values
+    var quantities = [
+        ["Uploaded by " + "..." + linkData.target_pk.substr(-config.node.hoverLabel.publicKeyCharacters), linkData.amount_up + " bytes"],
+        ["Downloaded by " + "..." + linkData.source_pk.substr(-config.node.hoverLabel.publicKeyCharacters), linkData.amount_down + " bytes"]
```
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: First concatenation is not necessary: "Uploaded by ..." +

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 23]
```diff
@@ -41,45 +42,102 @@ function drawNodes(svg, data, on_click) {
...
+            mouseOverNode(d, hoverInfoLabel);
+        })
+        .on("mousemove", function () {
+            hoverInfoLabel
+                .style("left", d3.event.pageX + "px")
+                .style("top", d3.event.pageY + "px");
+        })
+        .on("mouseout", function () {
+            d3.select(".hoverInfoLabel").select("table.hoverInfoTable").selectAll("tr").remove();
+            hoverInfoLabel.style("opacity", 0);
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Why can't you reuse the existing circle?

Comment (rkeulemans) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Because otherwise the text label inside the node overrules the event triggering the label to appear.

File: TriblerGUI/widgets/trustpage/js/drawing.js (hunk 20 above)
Comment (rkeulemans) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Look below. They are still there.
